File: 2022/day_02.cpp
//
// Created by Wollenschneider Luis on 02.12.22.
//

#include "../utils.hpp"
#include <list>

#define DAY "02"
#define YEAR "2022"

struct RPS {
    int opponent;
    int player;
};

#define TYPE RPS

std::list<TYPE> parse_input(const std::string& filepath) {
    // check if path exists
    std::list<TYPE> input;
    std::ifstream file(filepath);
    if (!file.is_open()) {
        std::cout << "File not found" << std::endl;
        return input;
    }
    std::string line;
    while (std::getline(file, line)) {
        RPS rps = {
            line[0] - 'A',
            line[2] - 'X'
        };
        input.push_back(rps);
    }
    file.close();
    return input;
}

int part1(const std::list<TYPE>& inp) {
    int score = 0;
    for (const RPS& rps : inp) {
        int round = rps.player + 1;
        if (rps.opponent == rps.player) {
            // draw
            round += 3;
        } else if (rps.player == (rps.opponent + 1) % 3) {
            // win
            round += 6;
        }
        score += round;
    }
    return score;
}

int part2(const std::list<TYPE>& inp) {
    int score = 0;
    for (const RPS& rps : inp) {
        int round = 0;
        switch (rps.player) {
            case 0:
                round = (rps.opponent - 1 + 3) % 3 + 1;
                break;
            case 1:
                round = 3;
                round += rps.opponent + 1;
                break;
            case 2:
                round = 6;
                round += (rps.opponent + 1) % 3 + 1;
                break;
            default:
                break;
        }
        score += round;
    }
    return score;
}

int main() {
    std::list<TYPE> inp = parse_input(YEAR "/inputs/day_" DAY ".txt");
    std::list<TYPE> test_inp = parse_input(YEAR "/tests/day_" DAY ".txt");

    int expected_result_part1 = get_expected_result<int>(YEAR "/tests/results/day_" DAY "_1.txt");

    int res = part1(test_inp);
    bool eval_res = evaluate_results(res, expected_result_part1);
    res = part1(inp);
    std::cout << "Part 1: " << res << std::endl;

    if (!eval_res) {
        return 1;
    }

    std::cout << std::endl;
    int expected_result_part2 = get_expected_result<int>(YEAR "/tests/results/day_" DAY "_2.txt");

    res = part2(test_inp);
    evaluate_results(res, expected_result_part2);
    res = part2(inp);
    std::cout << "Part 2: " << res << std::endl;

    return 0;
}

File: 2025/day_02.ts
import { solve } from '../utils';

const YEAR = 2025;
const DAY = 2;

function parseInput(input: string): any {
    return input.split(',').map(e => e.split('-').map(Number));
}


function valid_1(x: number): boolean {
    let s = x.toString();
    if (s.length % 2 === 1) return true;
    let mid = s.length / 2;
    let left = s.slice(0, mid);
    let right = s.slice(mid);
    return left !== right;
}

function part1(input: string): any | null {
    let inp = parseInput(input);

    let s = 0;
    for (let [a, b] of inp) {
        for (let i = a; i <= b; i++) {
            if (!valid_1(i)) {
                s += i;
            }
        }
    }
    return s;
}

function valid_2(x: number): boolean {
    let s = x.toString();
    for (let i = 1; i < s.length; i++) {
        if (s.length % i !== 0) continue;
        let v = false;
        for (let j = 0; j < s.length - i; j += i) {
            let part = s.slice(j, j + i);
            let next_part = s.slice(j + i, j + 2 * i);
            if (part !== next_part) {
                v = true;
                break;
            }
        }
        if (!v) return false;
    }
    return true;
}

function part2(input: string): any | null {
    let inp = parseInput(input);

    let s = 0;
    for (let [a, b] of inp) {
        for (let i = a; i <= b; i++) {
            if (!valid_2(i)) {
                s += i;
            }
        }
    }
    return s;
}

solve(part1, YEAR, DAY, 1);
solve(part2, YEAR, DAY, 2);